### PR TITLE
[SPARK-15590] [WEBUI] Paginate Job Table in Jobs tab

### PR DESCRIPTION
**Pull request by nblintao**
## What changes were proposed in this pull request?

This patch adds pagination support for the Job Tables in the Jobs tab. Pagination is provided for all of the three Job Tables (active, completed, and failed). Interactions (jumping, sorting, and setting page size) for paged tables are also included.

The diff didn't keep track of some lines based on the original ones. The function `makeRow`of the original `AllJobsPage.scala` is reused. They are separated at the beginning of the function `jobRow` (L427-439) and the function `row`(L594-618) in the new `AllJobsPage.scala`.
## How was this patch tested?

Tested manually by using checking the Web UI after completing and failing hundreds of jobs.
Generate completed jobs by:

``` scala
val d = sc.parallelize(Array(1,2,3,4,5))
for(i <- 1 to 255){ var b = d.collect() }
```

Generate failed jobs by calling the following code multiple times:

``` scala
var b = d.map(_/0).collect()
```

Interactions like jumping, sorting, and setting page size are all tested.

This shows the pagination for completed jobs:
paginate success jobs

This shows the sorting works in job tables:
sorting

This shows the pagination for failed jobs and the effect of jumping and setting page size:
paginate failed jobs

### CODE REVIEW

**Comment by andrewor14** [comment]
add to whitelist

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala [hunk 1]
```diff
@@ -369,3 +361,246 @@ private[ui] class AllJobsPage(parent: JobsTab) extends WebUIPage("") {
...
+
+  override def dataSize: Int = data.size
+
+  override def sliceData(from: Int, to: Int): Seq[JobTableRowData] = {
+    val r = data.slice(from, to)
+    _slicedJobIds = r.map(_.jobData.jobId).toSet
+    r
+  }
+
+  def slicedJobIds: Set[Int] = _slicedJobIds
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: used anywhere?

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala [hunk 2]
```diff
@@ -369,3 +361,246 @@ private[ui] class AllJobsPage(parent: JobsTab) extends WebUIPage("") {
...
+      ordering
+    }
+  }
+
+}
+private[ui] class JobPagedTable(
+    data: Seq[JobUIData],
+    jobTag: String,
+    basePath: String,
+    // TODO: Should use StageID instead of Int below, but cannot resolve. How?
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: what do you mean by this?

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala [hunk 3]
```diff
@@ -369,3 +361,246 @@ private[ui] class AllJobsPage(parent: JobsTab) extends WebUIPage("") {
...
+      s"&$jobTag.desc=$desc" +
+      s"&$pageSizeFormField=$pageSize"
+  }
+
+  override def goButtonFormPath: String = {
+    val encodedSortColumn = URLEncoder.encode(sortColumn, "UTF-8")
+    s"$basePath?$jobTag.sort=$encodedSortColumn&$jobTag.desc=$desc"
+  }
+
+  def headers: Seq[Node] = {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: is this used anywhere? I can't find it

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala [hunk 4]
```diff
@@ -369,3 +361,246 @@ private[ui] class AllJobsPage(parent: JobsTab) extends WebUIPage("") {
...
+              {header}
+            </th>
+          }
+        }
+      }
+    }
+    <thead>{headerRow}</thead>
+  }
+
+  def row(jobTableRow: JobTableRowData): Seq[Node] = {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: same here, is this used?

**Comment by andrewor14** [comment]
@nblintao this looks pretty good. There were a few methods that don't seem to be used anywhere. Am I missing something?

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala (hunk 1 above)
Comment (nblintao) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: Sorry, it's deprecated. I will remove it. Thanks.

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala (hunk 2 above)
Comment (nblintao) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: in the class `JobProgressListener`, the `stageIdToData` is defined as `HashMap[(StageId, StageAttemptId), StageUIData]` and the `stageIdToInfo` is defined as `HashMap[StageId, StageInfo]`.
But they cannot be resolved when the types like `StageId`, `StageAttemptId` are used. (They are defined in `JobProgressListener`, which I think could be used by others.) So I used `Int` here temporary.
I think this should be fixed before the patch is merged, but I don't know how for now.
Any suggestions?

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala (hunk 3 above)
Comment (nblintao) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: It implements the `headers` in the trait `PagedTable`. Shall I add a notion of `override` here to make it clear?

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala (hunk 4 above)
Comment (nblintao) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: Same here, it implements the `row` in the trait `PagedTable`.

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala (hunk 3 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: yes

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala (hunk 2 above)
Comment (nblintao) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: The todo is removed.

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala (hunk 3 above)
Comment (nblintao) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: added

**Comment by ajbozarth** [comment]
This looks great, I checked out the code and it looks good locally on my machine as well. I've read through the code and it looks good with the fixes @andrewor14 suggested so LGTM

**Comment by andrewor14** [comment]
also cc @zsxwing

**Comment by ajbozarth** [comment]
Just opened this in the history server to double check the bug I found on your stages pr and discovered that this is completely non-functional on the history server, so I'm striking my earlier ~~LGTM~~

**Comment by nblintao** [comment]
@ajbozarth Thanks for your checking and reporting. I didn't notice that the UI for history servers are different and just changed the AllJobsPage.

I will try to check the page for history servers. But do you think it would be better if we use another PR to solve this (adding the pagination support for the jobs page in history servers)?

**Comment by ajbozarth** [comment]
They actually share the core majority of their code so your pagination is already there, it just can't change pages or change the number of shown jobs. I think its has to do with how your calling and parsing the http requests since I didn't see any parameters in the url when I would try to change pages

**Comment by nblintao** [comment]
@ajbozarth Got it, that's a fatal bug. I will fix it in this PR. Thanks again!

**Comment by nblintao** [comment]
This PR has known bugs about linking. I have fixed the related bugs in #13708. In order to avoid conflicts, I will not update this PR until #13708 is merged.
So, please do NOT review or test this PR for now. It would be nice if you could help review or test #13708. Thanks!

**Comment by nblintao** [comment]
I believe this commit has resolved the bugs reported by @ajbozarth. It looks well on history server pages now, and it could keep the status of other tables while changing one.
Could you please help test or review it if you are available? Thanks! @andrewor14 @zsxwing @ajbozarth

**Comment by ajbozarth** [comment]
@nblintao I just finished testing your changes and it looks good. I did find a odd behavior though, but I think it's intentional in the original pagination code. If you make the table large enough to fit all the jobs the pagination toolbar disappears. Given you can just click the back button to fix it I don't think it's an issue to be fixed, but it is a bit annoying. Another fyi is that only ~1000 jobs are shown still, but I'm assuming the pagination doesn't fix the reasoning behind only showing the most recent 1000 jobs.

tl;dr LGTM

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala [hunk 5]
```diff
@@ -210,64 +214,69 @@ private[ui] class AllJobsPage(parent: JobsTab) extends WebUIPage("") {
...
   }
 
-  private def jobsTable(jobs: Seq[JobUIData]): Seq[Node] = {
+  private def jobsTable(
+      request: HttpServletRequest,
+      jobTag: String,
+      jobs: Seq[JobUIData]): Seq[Node] = {
+    val allParameters = request.getParameterMap.asScala.toMap
+    val parameterOtherTable = allParameters.filterNot(_._1.startsWith(jobTag))
+      .map(para => para._1 + "=" + para._2(0))
```
Comment (yhuai) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: What does `parameterOtherTable` do?

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala [hunk 6]
```diff
@@ -210,64 +214,69 @@ private[ui] class AllJobsPage(parent: JobsTab) extends WebUIPage("") {
...
+        parent.jobProgresslistener.stageIdToInfo,
+        parent.jobProgresslistener.stageIdToData,
+        currentTime,
+        jobIdTitle,
+        pageSize = jobPageSize,
+        sortColumn = jobSortColumn,
+        desc = jobSortDesc
+      ).table(page)
+    } catch {
+      case e @ (_ : IllegalArgumentException | _ : IndexOutOfBoundsException) =>
```
Comment (yhuai) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: When will this these two exceptions get triggered?

**Comment by yhuai** [comment]
@nblintao Can you comment on your PR to explain which parts are new code and which parts are based on existing code?

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala [hunk 7]
```diff
@@ -369,3 +375,246 @@ private[ui] class AllJobsPage(parent: JobsTab) extends WebUIPage("") {
...
+        }
+      }
+    }
+    <thead>{headerRow}</thead>
+  }
+
+  override def row(jobTableRow: JobTableRowData): Seq[Node] = {
+    val job = jobTableRow.jobData
+
+    <tr id={"job-" + job.jobId}>
```
Comment (yhuai) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: Copied from https://github.com/apache/spark/pull/13620/files#diff-9a486b2f4e30d5375c3541571aec616cL240?

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala (hunk 5 above)
Comment (nblintao) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: It keeps the status of other tables on the page. See https://github.com/apache/spark/pull/13708#issuecomment-226642404

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala (hunk 7 above)
Comment (nblintao) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: Yes.

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala (hunk 6 above)
Comment (nblintao) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: It was following @zsxwing's code for `TaskPagedTable` (https://github.com/apache/spark/commit/4f7f1ee378e80b33686508d56e133fc25dec5316#diff-fa4cfb2cce1b925f55f41f2dfa8c8501R265). I haven't found concrete positions where they will be triggered yet.
But I think catching exceptions here is necessary, because I'm not sure my code for `JobPagedTable` is robust enough without exceptions. Since this exception information is shown in the UI, the user could easily notice which table cannot be rendered.

**Comment by nblintao** [comment]
@yhuai Updated:

> The diff didn't keep track of some lines based on the original ones. The function `makeRow`of the original `AllJobsPage.scala` is reused. They are separated at the beginning of the function `jobRow` (L427-439) and the function `row`(L594-618) in the new `AllJobsPage.scala`.

Thanks for your review!

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala (hunk 6 above)
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: > When will this these two exceptions get triggered?

As the parameters are from the user input (the url), the user may type some illegal paramters in the url, which causes these exceptions.

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
LGTM. Let's run the test again since this is a pretty old PR. Thanks!

**Comment by zsxwing** [comment]
Merging to master.

**Comment by nblintao** [comment]
Great! Thanks, @andrewor14 @ajbozarth @yhuai @zsxwing.
